### PR TITLE
Refactoring of ILL Reflectometry Suite

### PR DESCRIPTION
**Pull request by gvardany**
**Description of work.**

This PR refactors the reflectometry reduction algorithms for ILL.
Few redundancies are removed, the consistency of the results is greatly improved wrt cosmos.
Because of few fundamental changes (such as removal of the input workspace in preprocess, rewrite of DAN calibration, foreground fitting in loader, etc.), the tests also needed a major rewrite.
This perhaps reduces the unit test coverage, but time permitting they will be completed in future.

**To test:**
Results are more consistent with Lamp/Cosmos, and captured by the system tests, so code review is all you can do.

Part of #27197.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/algorithms/ReflectometryILLPreprocess-v1.rst [hunk 1]
```diff
@@ -25,103 +24,47 @@ The algorithm can be thought as an 'advanced loader', and should be used to load
...
 
-Background, on the other hand, is a set of pixels which are be used for fitting a constant or linear background by :ref:`CalculatePolynomialBackground <algm-CalculatePolynomialBackground>`.
-
-The foreground pixels are defined by the foreground centre and *ForegroundHalfWidth* property. In normal use cases, the foreground center (workspace index) is taken from the fitting in :ref:`LoadILLReflectometry <algm-LoadILLReflectometry>`. This can be overridden by giving the pixel as *BeamCentre*. Fractional values are rounded to nearest integer. The full process of deciding the foreground centre is as follows:
+Background, on the other hand, is a set of pixels which are be used for average (default), fitted constant or linear background by :ref:`CalculatePolynomialBackground <algm-CalculatePolynomialBackground>`.
```
Comment (gemmaguest) [inline docs/source/algorithms/ReflectometryILLPreprocess-v1.rst:52]: Should the average background calculation method be mentioned here?

File: docs/source/algorithms/SetMDFrame-v1.rst [hunk 2]
```diff
@@ -30,11 +30,11 @@ Usage
 
 **Example - Sample usage of SetMDFrame**
 
-.. testcode:: SetMDFrameExample
+.. code-block:: python
```
Comment (gemmaguest) [inline docs/source/algorithms/SetMDFrame-v1.rst:33]: I'm not sure why this file is changed?

File: docs/source/algorithms/SofTwoThetaTOF-v1.rst [hunk 3]
```diff
@@ -30,8 +30,7 @@ Usage
 
 **Example - Applying SofTwoThetaTOF and plotting the result**
 
-.. plot::
-   :include-source:
+.. code-block:: python
```
Comment (gemmaguest) [inline docs/source/algorithms/SofTwoThetaTOF-v1.rst:33]: I'm not sure why this file is changed?

**Review by gemmaguest — CHANGES_REQUESTED** [review]
Very nice code and these all look like sensible changes :+1: Just a few minor queries inline.

As requested I have mainly just done a code review. Test coverage looks sufficient and validation will be down to the ILL. I ran a few usage examples too and they ran ok.

It's not related to this PR, but I noticed when trying to run the usage examples that the data is not included in the UsageData download. I can see that it is in the unit test data so I managed to get them to run, but users will probably not have this. Could they be added in the future?

File: docs/source/algorithms/ReflectometryILLPreprocess-v1.rst (hunk 1 above)
Comment (gvardany) [reply docs/source/algorithms/ReflectometryILLPreprocess-v1.rst:52]: Yes, that's what I refer to as default in parentheses, perhaps could have been formulated better.

File: docs/source/algorithms/SetMDFrame-v1.rst (hunk 2 above)
Comment (gvardany) [reply docs/source/algorithms/SetMDFrame-v1.rst:33]: Was getting systematic segfaults on building the documentation, I can revert this but there are genuine problems with this and the SofTwoThetaTOF doc tests.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLPreprocess.py [hunk 4]
```diff
@@ -543,6 +466,24 @@ def _normaliseToSlits(self, ws):
...
+            condition = (((x >= ranges[0]) & (x <= ranges[1])) | ((x >= ranges[2]) & (x <= ranges[3])))
+        bkg_region = np.extract(condition, y)
+        bkg_region = bkg_region.reshape((nspec, int(bkg_region.size/nspec)))
+        bkg = np.mean(bkg_region, axis=1)
+        for channel in range(nspec):
+            transposedBkgWS.setE(channel, np.zeros(blocksize))
+            transposedBkgWS.setY(channel, bkg[channel] * np.ones(blocksize))
+        return transposedBkgWS
+
     def _subtractFlatBkg(self, ws):
```
[the remark this answers is not in the document]
Comment (gvardany) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLPreprocess.py:487]: This should be possible, I'll make note in the backlog for the future.

File: docs/source/algorithms/ReflectometryILLPreprocess-v1.rst (hunk 1 above)
Comment (gemmaguest) [reply docs/source/algorithms/ReflectometryILLPreprocess-v1.rst:52]: Ah yes, it reads fine this morning. Must just have been tired last night!

File: docs/source/algorithms/SetMDFrame-v1.rst (hunk 2 above)
Comment (gemmaguest) [reply docs/source/algorithms/SetMDFrame-v1.rst]: Ah ok. No problem, I just wanted to understand :)

**Comment by gvardany** [comment]
Thanks for the review. I have committed the fixes. As for the usage data, I think your comment is valid for almost all of our reductions. Is there a way to automatically include all the test data in usage examples?

**Comment by gemmaguest** [comment]
> 
> 
> Thanks for the review. I have committed the fixes. As for the usage data, I think your comment is valid for almost all of our reductions. Is there a way to automatically include all the test data in usage examples?

I don't think so. I think you have to manually add them to the UsageData.zip file on sourceforge...

**Comment by gvardany** [comment]
I think we should add all our usage example data to the .zip as a maintenance task.

**Comment by gvardany** [comment]
That should be all I guess.

**Review by gemmaguest — APPROVED** [review]
Thanks for the changes, this all looks good to me now :+1:

**Review by DavidFair — APPROVED** [review]
I'm happy with this and there's a good work here especially around the docs by removing some implementation details users don't care about in favour of more user-centric information.

**Pull request merged**